docs/spec/overload.rst

```diff
@@ -99,8 +109,349 @@ variable is not sufficient.
+Overload definitions within stub files, protocols, and on abstract methods
+within abstract base classes are exempt from this check.
+
+If one overload signature is decorated with ``@staticmethod`` or
+``@classmethod``, all overload signatures must be similarly decorated. The
+implementation, if present, must also have a consistent decorator. Type
+checkers should report an error if these conditions are not met.
+
+If a ``@final`` or ``@override`` decorator is supplied for a function with
+overloads, the decorator should be applied only to the overload implementation
+if it is present. If an overload implementation isn't present (for example, in
+a stub file), the ``@final`` or ``@override`` decorator should be applied only
+to the first overload. Type checkers should enforce these rules and generate
+an error when they are violated. If a ``@final`` or ``@override`` decorator
+follows these rules, a type checker should treat the decorator as if it is
+present on all overloads.
+
+Overloads are allowed to use a mixture of ``async def`` and ``def`` statements
+within the same overload definition. Type checkers should convert
+``async def`` statements to a non-async signature (wrapping the return
+type in a ``Coroutine``) before testing for implementation consistency
+and overlapping overloads (described below).
+
+
+Implementation consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If an overload implementation is defined, type checkers should validate
+that it is consistent with all of its associated overload signatures.
+The implementation should accept all potential sets of arguments
+that are accepted by the overloads and should produce all potential return
+types produced by the overloads. In typing terms, this means the input
+signature of the implementation should be :term:<assignable> to the input
+signatures of all overloads, and the return type of all overloads should be
+assignable to the return type of the implementation.
+
+If the implementation is inconsistent with its overloads, a type checker
+should report an error::
+
+  @overload
+  def func(x: str, /) -> str: ...
+  @overload
+  def func(x: int) -> int: ...
+
+  # This implementation is inconsistent with the second overload
+  # because it does not accept a keyword argument ``x`` and the
+  # the overload's return type ``int`` is not assignable to the
+  implementation's return type ``str``.
+  def func(x: int | str, /) -> str:
+    return str(x)
+
+When a type checker checks the implementation for consistency with overloads,
+it should first apply any transforms that change the effective type of the
+implementation including the presence of a ``yield`` statement in the
+implementation body, the use of ``async def``, and the presence of additional
+decorators.
+
+
+Overlapping overloads
+^^^^^^^^^^^^^^^^^^^^^
+
+If two overloads can accept the same set of arguments, they are said
+to "partially overlap". If two overloads partially overlap, the return type
```

Comment:
@JelleZijlstra, @JukkaL, I want to confirm that you are OK with the proposed definition of "partially overlapping overloads" in light of the response @ilevkivskyi posted to [this mypy issue](https://github.com/python/mypy/issues/18581). I think Ivan's point is worth considering, although I'm not sure how to write the specification for "cases that could fool the type checker into choosing the wrong overload". Intuitively, I know what he means here, but a precise spec might be difficult to formulate. It looks from [this historical comment](https://github.com/python/mypy/issues/5119#issuecomment-393221830) that mypy contains quite a few rules and various exceptions to those rules.

Reply:
How does pyright currently check overlapping overloads?

The text already contains a number of special cases (`__get__`, ignoring multiple inheritance, etc.). Also, it seems to me that overlap is a somewhat isolated issue; the rules we set out for evaluating overload calls still work even if some overloads overlap. It might be better then to leave the spec vague, saying that type checkers may choose to flag some unsafe overlaps, without trying to come up with precise rules.

Reply:
> How does pyright currently check overlapping overloads?

The implementation is complicated. It leverages assignability rules but internally sets a flag that changes some of the normal assignability rules. For example, normally `Any` is assignable to every other type, but when this flag is set, `Any` is not assignable to any type (even itself). There are a number of other changes in the algorithm, but they attempt to implement the behavior described in this PR. 

If we don't have agreement on the desired behavior, then I agree that it's best to keep the spec loose here. Actually, I'm not sure it makes sense to discuss it at all in the spec if we are not in agreement on the intended behavior. So maybe we should just delete that entire section and the corresponding conformance tests.

Reply:
I've deleted the section of the spec related to overlapping overloads since we weren't able to achieve consensus on this topic.